### PR TITLE
Stm32h755 soc support and nucleo-h755zi-q support

### PR DESCRIPTION
**Pull request by mbanducci-sandc**
This PR adds support for the stm32h755 soc and the ST nucleo-h755zi-q development board. 

The h755 is a close relative of the h745 which adds cryptography and hash accelerator peripherals.

The nucleo-h755zi-q is a nucleo-144 development board for the stm32h755. It shares the same layout as other, similar nucleo-144 boards such as the nucleo-h745zi-q already supported by Zephyr.

Thanks!

### CODE REVIEW

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+.. note::
+
+   Check if your ST-LINK V3 has newest FW version. It can be done with `STM32CubeIDE`_
+
+Flashing
+========
+
+Nucleo H755ZI-Q board includes an ST-LINK/V3 embedded debug tool interface.
+
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/doc/index.rst:193]:
```

The board is configured to be flashed using west `STM32CubeProgrammer`_ runner
for both cores, so its installation is required to be able to flash the board.
The target core is detected automatically.

It is advised to use `STM32CubeProgrammer`_ to check and update option bytes
configuration and flash ``nucleo_h755zi_q/stm32h755xx/m7`` and
``nucleo_h755zi_q/stm32h755xx/m4`` targets.

```

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+Flashing operation will depend on the target to be flashed and the SoC
+option bytes configuration.
+
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/doc/index.rst:196]:
```

It is advised to use `STM32CubeProgrammer`_ to check and update option bytes
configuration.

```

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+By default:
+
+  - CPU0 (Cortex-M7) boot address is set to 0x80000000 (OB: BOOT_CM7_ADD0)
+  - CPU1 (Cortex-M4) boot address is set to 0x81000000 (OB: BOOT_CM4_ADD0)
+
+Also, default out of the box board configuration enables CM7 and CM4 boot when
+board is powered (Option bytes BCM7 and BCM4 are checked).
+In that configuration, Kconfig boot option ``STM32H7_BOOT_CM4_CM7`` should be selected.
+Zephyr flash configuration has been set to meet these default settings.
+
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/doc/index.rst:206]:
```

Alternatively, openocd or JLink can also be used to flash the board using
the ``--runner`` (or ``-r``) option:

.. code-block:: console

   $ west flash --runner openocd
   $ west flash --runner jlink
```

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+   https://www.st.com/resource/en/reference_manual/dm00176879-stm32h745755-and-stm32h747757-advanced-armbased-32bit-mcus-stmicroelectronics.pdf
+
+.. _OpenOCD installing Debug Version:
+   https://github.com/zephyrproject-rtos/openocd
+
+.. _OpenOCD installing with ST-LINK V3 support:
+   https://mbd.kleier.net/integrating-st-link-v3.html
+
+.. _STM32CubeIDE:
+   https://www.st.com/en/development-tools/stm32cubeide.html
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/doc/index.rst:293]:
```

.. _STM32CubeProgrammer:
   https://www.st.com/en/development-tools/stm32cubeprog.html
```

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q.dtsi [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2019 Linaro Limited
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q.dtsi:2]: copyright

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m4_defconfig [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m4_defconfig:2]: copyright

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7.dts [hunk 7]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7.dts:2]: copyright

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig:2]: copyright

File: boards/st/nucleo_h755zi_q/support/openocd.cfg [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# STM32H745ZI Nucleo board OpenOCD ST-LINK V3 configuration
+#
+# Copyright (c) 2020 Alexander Kozhinov  <AlexanderKozhinov@yandex.com>
```
Comment (ajarmouni-st) [inline boards/st/nucleo_h755zi_q/support/openocd.cfg:3]: copyright

File: samples/drivers/flash_shell/sample.yaml
Comment (ajarmouni-st) [inline samples/drivers/flash_shell/sample.yaml:1]: should be in a different commit

**Review by ajarmouni-st — DISMISSED** [review]
Thanks for working on this, here is a first batch of comments.

File: boards/st/nucleo_h755zi_q/board.cmake [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
+board_runner_args(jlink "--device=STM32H755ZI" "--speed=3300")
+if(CONFIG_BOARD_NUCLEO_H755ZI_Q_STM32H755XX_M7)
+board_runner_args(openocd --target-handle=_CHIPNAME.cpu0)
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/board.cmake:6]: 2 space indent

File: boards/st/nucleo_h755zi_q/board.cmake [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+elseif(CONFIG_BOARD_NUCLEO_H755ZI_Q_STM32H755XX_M4)
+board_runner_args(openocd --target-handle=_CHIPNAME.cpu1)
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/board.cmake:8]: and here

File: boards/st/nucleo_h755zi_q/doc/img/nucleo_h755zi_q.jpg
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/doc/img/nucleo_h755zi_q.jpg:1]: convert to webp, then put through https://tinypng.com/

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+
+Nucleo H755ZI-Q board includes an ST-LINK/V3 embedded debug tool interface.
+
+The board is configured to be flashed using west `STM32CubeProgrammer`_ runner
+for both cores, so its installation is required to be able to flash the board.
+The target core is detected automatically.
+
+It is advised to use `STM32CubeProgrammer`_ to check and update option bytes
+configuration and flash ``nucleo_h755zi_q/stm32h755xx/m7`` and
+``nucleo_h755zi_q/stm32h755xx/m4`` targets.
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/doc/index.rst:201]: *board targets

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m4_defconfig [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+# By default SERIAL peripherals are assigned to m7
+
+# Enable uart driver
+#CONFIG_SERIAL=y
+
+# Console
+#CONFIG_CONSOLE=y
+#CONFIG_UART_CONSOLE=y
+
+# enable pin controller
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m4_defconfig:19]: *Enable

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable Clock
+CONFIG_CLOCK_CONTROL=y
+
+# enable pin controller
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig:26]: *Enable

File: boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# Enable HW stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# Enable UART ( disable to assign to M4 core)
+CONFIG_SERIAL=y
+
+# Console ( disable to assign to M4 core)
```
Comment (nordicjm) [inline boards/st/nucleo_h755zi_q/nucleo_h755zi_q_stm32h755xx_m7_defconfig:16]: remove space after (

**Review by gautierg-st — DISMISSED** [review]

**Comment by mkaricheri** [comment]
Would be great if we can see few more approval for this pull request. Thanks

**Review by marwaiehm-st — DISMISSED** [review]

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+for both cores, so its installation is required to be able to flash the board.
+The target core is detected automatically.
+
+It is advised to use `STM32CubeProgrammer`_ to check and update option bytes
+configuration and flash ``nucleo_h755zi_q/stm32h755xx/m7`` and
+``nucleo_h755zi_q/stm32h755xx/m4`` board targets.
+
+By default:
+
+  - CPU0 (Cortex-M7) boot address is set to 0x80000000 (OB: BOOT_CM7_ADD0)
```
Comment (FRASTM) [inline boards/st/nucleo_h755zi_q/doc/index.rst:191]:
```suggestion
  - CPU0 (Cortex-M7) boot address is set to 0x08000000 (OB: BOOT_CM7_ADD0)
```
(this is also wrong for the stm32h747i_disco)

File: boards/st/nucleo_h755zi_q/doc/index.rst [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+  - CPU1 (Cortex-M4) boot address is set to 0x81000000 (OB: BOOT_CM4_ADD0)
```
Comment (FRASTM) [inline boards/st/nucleo_h755zi_q/doc/index.rst:192]:
```suggestion
  - CPU1 (Cortex-M4) boot address is set to 0x08100000 (OB: BOOT_CM4_ADD0)
```
(this is also wrong for the stm32h747i_disco)

**Review by FRASTM — COMMENTED** [review]
LGTM

**Review by FRASTM — APPROVED** [review]

**Review by marwaiehm-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by mkaricheri — APPROVED** [review]
Looks good to me

**Comment by mkaricheri** [comment]
@erwango we are Curious as to why this is still waiting to be merged? Is there anything to be done at our end? Thanks
